add pg DOMAIN type reflection

ovangle opened the pull request:
<!-- Provide a general summary of your proposed changes in the Title field above -->

### Description
<!-- Describe your changes in detail -->

- Adds a new method `PGDialect._reflect_type(format_type: str, type_description: Optional[str] = None) -> TypeEngine[Any]` to the `PGDialect` class.
- Ensure that `DOMAIN` column types are returned as the `DOMAIN` schema type when reflected, rather than returning the basis type. 

### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

-->

This pull request is:

- [ ] A documentation / typographical / small typing error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

Commit:
Fix #10693: postgresql dialect should reflect DOMAIN types

Also, refactor postgresql dialect type reflection into new protected
dialect instance method `PGDialect._reflect_type`

        PGDialect._reflect_type(format_type, enums, domains, type_description)

test #10693:

- psycopg and pg8000 raise different dbapi exceptions when a domain
  constraint fails. Weaken test assertion to catch both

review #10693:
- type_description is required parameter to _reflect_type
- fix docstring typo
- make arrayspec regex a class variable
- Use equality checks on attype_args
- reflected array dimensionality is always 1

CaselIT commented:
Hi, thanks for the pr

> Adds a new method PGDialect._reflect_type(format_type: str, type_description: Optional[str] = None) -> TypeEngine[Any] to the PGDialect class.

would have been easier to look what changed without this change. Other than the name it should be ok to keep though

ovangle commented:
That's kinda the first thing I did -- `get_column_info` was serving two purposes

- parsing the `format_type` into the corresponding schema type object (which I moved into `_reflect_type`); and
- rectonstructing the column itself

So the previous domain code had to do some weird control flow stuff to attempt to recurse on just the "reflect the schema type" part of it...

Names are obviously whatever you think they should be.

CaselIT commented:
Yes I agree, this organisation is better

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 67fa39415e09b83501adc3fa3a63c6c03032b85a of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change 67fa39415e09b83501adc3fa3a63c6c03032b85a: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

ovangle commented:
Apologies for test suite failures. I did run full suites at one point, but as usual I touched things up for "readability"...

This time entire base and `--db postgresql` suites were run to completion, because I'm not relying on CI to if someone has to manually trigger it each time.

I also rebased from main (my commits will eventually still need to be squashed though).

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 582f0c79da13f18400c3cb62c3154ab5c2a67710 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 582f0c79da13f18400c3cb62c3154ab5c2a67710 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

ovangle commented:
😳 I'll have to configure format on save in this workspace. `--db pg8000` should be running to completion now too.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 1b5c248bd7d929ee370376af0b7711319c311d8f of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 1b5c248bd7d929ee370376af0b7711319c311d8f added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented:
**Michael Bayer** (zzzeek) wrote:

this needs a changelog file and one of us here will also create a 2.0 backport

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented:
**Michael Bayer** (zzzeek) wrote:

wow great work , there is a lot here

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3643,20 +3643,152 @@ def get_multi_columns(
+        format_type: Optional[str],
+        domains: dict[str, ReflectedDomain],
+        enums: dict[str, ReflectedEnum],
+        type_description: Optional[str] = None,
```
**Federico Caselli** (CaselIT) wrote:

this is always provided so I don't think there is any reason to make it an optional str

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

ovangle replied:
It's always provided, but is unused except to add additional information to errors. The function could technically be called without it. It could arguably default to `"<unknown source>"` or something.

CaselIT replied:
it's an internal function likely ever only called by a single place, so it's probably fine to make it required

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3643,20 +3643,152 @@ def get_multi_columns(
+        on the information available in the format_type.
+
+        If the `format_type` cannot be associated with a known `ischema_names`,
+        then it is treated it is treated a reference to a known PostgreSQL
```
**Federico Caselli** (CaselIT) wrote:

typo: `then it is treated it is treated` -> `then it is treated`

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3643,20 +3643,152 @@ def get_multi_columns(
+
+        attype_args_match = self._format_type_args_pattern.search(format_type)
+        if attype_args_match and attype_args_match.group(1):
+            attype_args = re.split(r"\s*,\s*", attype_args_match.group(1))
```
**Federico Caselli** (CaselIT) wrote:

would make sense to make also this regexp a class variable

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3643,20 +3643,152 @@ def get_multi_columns(
+        args, kwargs = (), {}
+
+        if attype == "numeric":
+            if len(attype_args) >= 2:
```
**Federico Caselli** (CaselIT) wrote:

==

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3643,20 +3643,152 @@ def get_multi_columns(
+
+        elif attype in ("timestamp with time zone", "time with time zone"):
+            kwargs["timezone"] = True
+            if len(attype_args) >= 1:
```
**Federico Caselli** (CaselIT) wrote:

==

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3643,20 +3643,152 @@ def get_multi_columns(
+            "time",
+        ):
+            kwargs["timezone"] = False
+            if len(attype_args) >= 1:
```
**Federico Caselli** (CaselIT) wrote:

==

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3643,20 +3643,152 @@ def get_multi_columns(
+
+        elif attype == "bit varying":
+            kwargs["varying"] = True
+            if len(attype_args) >= 1:
```
**Federico Caselli** (CaselIT) wrote:

==

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3643,20 +3643,152 @@ def get_multi_columns(
+            if field_match:
+                kwargs["fields"] = field_match.group(1)
+
+            if len(attype_args) >= 1:
```
**Federico Caselli** (CaselIT) wrote:

==

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3643,20 +3643,152 @@ def get_multi_columns(
+                args = (domain["name"], data_type)
+
+                # FIXME: collation not reflected
+                # kwargs["collation"] = "???"
```
**Federico Caselli** (CaselIT) wrote:

seems like something that could be reflected by changing the domain reflection query.

I can look into it but it's unlikely to be very important so it's probably fine to leave as todo

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3643,20 +3643,152 @@ def get_multi_columns(
+            return sqltypes.NULLTYPE
+
+        data_type = schema_type(*args, **kwargs)
+        for _ in range(array_dim):
```
**Federico Caselli** (CaselIT) wrote:

why this change? it's unrelated and there are no test for array_dim > 1

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

ovangle replied:
You're right, original code probably didn't work for multi-dimensional arrays too.

CaselIT replied:
I mean we can keep it, but we would need a test for it, to verify that it actually works

ovangle replied:
Sorry for delay, I've been busy with other stuff. After some research, [dimensionality of array types is not enforced by postgres](https://www.postgresql.org/docs/current/arrays.html), so...

I'm guessing this should just be `if array_dim >= 1:` instead?

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3667,131 +3799,26 @@ def _handle_array_type(attype):
+            )
+
             default = row_dict["default"]
+            if isinstance(coltype, DOMAIN) and not default:
```
**Federico Caselli** (CaselIT) wrote:

maybe _reflect_type could return `is_domain` too? to avoid the isinstance here and below?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

ovangle replied:
Would you prefer a `hasattr` check here and below?

CaselIT replied:
it's likely fine as is, just mentioned that since `_reflect_type` knows when it's returning a doman it could return that to the caller

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

change seems mostly ok. I've left a few more comments

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

CaselIT commented:
@ovangle just fyi, if you prefer that we take over the PR just say so. In any case thanks for the effort so far!

ovangle commented:
@CaselIT I don't mind either way. I'd be happy to make the requested changes, but I'm busy with other work.

You're welcome :D

CaselIT commented:
Sounds good, thanks!

ovangle commented:
OK, I've addressed all your comments and am just running the full test suite. I'll rebase from main and bundle everything into a single commit after these tests pass.

Regarding updating changelogs etc. or any further changes, I'm fine leaving that stuff to you if you still want to take over, @CaselIT

CaselIT commented:
Great, thanks!

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision adac164d191138265ecd64a28be91254a53a9c25 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset adac164d191138265ecd64a28be91254a53a9c25 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3661,20 +3661,158 @@ def get_multi_columns(
 
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3661,20 +3661,158 @@ def get_multi_columns(
+        domains: dict[str, ReflectedDomain],
+        enums: dict[str, ReflectedEnum],
+        type_description: str,
+    ) -> sqltypes.TypeEngine[Any]:
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3661,20 +3661,158 @@ def get_multi_columns(
+            )
+            return sqltypes.NULLTYPE
+
+        attype_args_match = self._format_type_args_pattern.search(format_type)
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3661,20 +3661,158 @@ def get_multi_columns(
+        # Each "[]" in array specs corresponds to an array dimension
+        array_dim = len(match_array_dim.group(1) or "") // 2
+
+        # Remove all parameters and array specs from format_type to obtain an
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3661,20 +3661,158 @@ def get_multi_columns(
+        args, kwargs = (), {}
+
+        print("attype", attype, "schema_type", schema_type)
+
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3661,20 +3661,158 @@ def get_multi_columns(
+            if len(attype_args) == 2:
+                precision, scale = map(int, attype_args)
+                args = (precision, scale)
+
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3661,20 +3661,158 @@ def get_multi_columns(
+            kwargs["timezone"] = True
+            if len(attype_args) == 1:
+                kwargs["precision"] = int(attype_args[0])
+
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3661,20 +3661,158 @@ def get_multi_columns(
+
+                if domain["constraints"]:
+                    # We only support a single constraint
+                    check_constraint = domain["constraints"][0]
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

* lib/sqlalchemy/dialects/postgresql/base.py (line 3654): Done

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3661,20 +3661,158 @@ def get_multi_columns(
+
+        data_type = schema_type(*args, **kwargs)
+        if array_dim >= 1:
+            # postgres does not preserve dimensionality or size of array types.
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

* doc/build/changelog/unreleased_21/10693.rst (line 1): I've placed this in 2.1 but I'm not against having it on 2.0 since I think the changes are quite minor.

What's your opinion mike? (cc @zzzeek)

CaselIT commented:
asked mike in chat, will backport to 2.0

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

* doc/build/changelog/unreleased_21/10693.rst (line 1): backported to 2.0

sqla-tester reviewed (commented):
**David Baumgold** (singingwolfboy) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

* doc/build/changelog/unreleased_21/10693.rst (line 10): Typo: "curtesy" -> "courtesy"

sqla-tester commented:
**Federico Caselli** (CaselIT) wrote:

indeed

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5239

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030

* doc/build/changelog/unreleased_21/10693.rst (line 10): Done

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030 has been **merged**. Congratulations! :)

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5239 has been **merged**. Congratulations! :)
